### ISSUE TITLE
Try to get loudness from ReplayGain tags

### ISSUE THREAD

**Issue by diogocp** [andrewrk/libgroove#103]
Most of my files already have ReplayGain tags, but they seem to be ignored when I play them in Groove Basin. Instead of just reading a few bytes of metadata, it re-scans the file. This obviously takes much longer and also soaks up a lot of CPU time.

For files with no ReplayGain tags, scanning would proceed as usual. But while we're at it, why not (optionally) write those tags afterwards?

**Comment by andrewrk** [comment]
Groove Basin distrusts ReplayGain tags written by third party software. One of the features of Groove Basin is that the same code that is used for loudness scanning is the same code that is used for playback, so you don't get weird glitches or mismatched calibrations. The downside of soaking up CPU time is notable; however I think we want to scan every song *anyway* for these reasons:

 * Determining exact song duration in case tags are incorrect about duration (sadly I see this very often)
 * Generating an acoustid fingerprint so that we can query acoustid.org and auto determine tags even when tags are unknown. Currently in groove basin we do compute this finger print but the feature to utilize it has not been implemented yet.
 * (possible feature) Determine BPM or other qualities about the song to potentially inform Auto DJ.
 * (planned) Compute song waveform to display in the seek bar.

Writing tags is planned, currently blocking on #94

**Comment by diogocp** [comment]
So, the first problem is that there is no way to tell which algorithm was used to calculate the ReplayGain values. This seems easy enough to solve with a replaygain_version tag. Still, the difference between the two should not be that noticeable.

Then you pointed out other reasons to scan the files. But does loudness scanning help with any of those things? I think they need to be considered separately. In fact, this reminds me I need to open another issue...

**Comment by thejoshwolfe** [comment]
If CPU usage is a problem, there are some related issues being discussed in https://github.com/andrewrk/groovebasin/issues/194 . I personally prefer to just reduce the CPU intensity of what Groove Basin is doing rather than change its trust strategy.

Back in my Winamp days, I suffered from bogus replaygain tags causing certain songs to be played much too quietly. Trusting third-party tags can go wrong, but scanning the song yourself always works. I'd rather not have two different ways of getting replaygain info, and I don't want to trust third-party tags.

> So, the first problem is that there is no way to tell which algorithm was used to calculate the ReplayGain values.

I'm glad you brought this up. Groove Basin actually doesn't use a "ReplayGain" algorithm; it uses a "loudness normalization" algorithm: [EBU R128](https://tech.ebu.ch/docs/r/r128.pdf). Any ReplayGain tags would be useless to Groove Basin, since that's the wrong algorithm. I suppose perhaps there could be some kind of conversion logic to switch between the two modes, but that sounds much too complicated and error prone. Compare that idea with simply always doing all the scanning ourselves and being certain that it's always correct and consistent.

> But does loudness scanning help with any of those things?

The loudness isn't what's useful for those other things. The process of scanning a song decodes the compressed audio to do processing on raw audio samples. Loudness is one analysis. The other analyses are also done with the decompressed audio data. I haven't done any measurements myself, but I expect CPU demand for the loudness analysis is very small compared with the CPU demand for decoding the compressed data into the raw samples. (And certainly the CPU demand for measuring the song length is very, very small compared with the decoding effort.)  Doing all the analyses separately would mean decoding the audio data several times, which I think would be wasteful. If we want to do one of the analyses, we should probably just go ahead and do them all while we're at it.

In conclusion, I really think https://github.com/andrewrk/groovebasin/issues/194 is the solution to this issue.

**Comment by diogocp** [comment]
> I haven't done any measurements myself, but I expect CPU demand for the loudness analysis is very small compared with the CPU demand for decoding the compressed data into the raw samples.

**Comment by diogocp** [comment]
> Trusting third-party tags can go wrong, but scanning the song yourself always works.

I do scan every file myself! Picard makes it easy. What I don't understand is why people keep bad metadata in their files and then complain about it. If you are not going to fix it, at least remove it! Picard or beets can remove all ReplayGain tags from your library in 5 seconds.

> I'm glad you brought this up. Groove Basin actually doesn't use a "ReplayGain" algorithm; it uses a "loudness normalization" algorithm: EBU R128.

ReplayGain 2.0 uses the same algorithm as EBU R128, that's why I mentioned a version tag. But honestly I doubt I could even tell the difference between 1.0 and R128. They both measure loudness, R128 just does it slightly better.

I plan to rescan my files with the EBU R128 algorithm, anyway, because why not? But I would much rather do it all at once than unpredictably at play time.

andrewrk/groovebasin#302 mentions triggering a full scan manually. That would sort of work for me if it included loudness scanning, although the idea that rescanning my library could take more than 3 days does make me itchy. Storing it in the files just makes a lot more sense, at least for me.

> I haven't done any measurements myself, but I expect CPU demand for the loudness analysis is very small compared with the CPU demand for decoding the compressed data into the raw samples.

Decoding has very little impact. Here's a quick benchmark of a 44MB FLAC file:

- FLAC, no true peak: 7.162s
- WAV, no true peak: 6.309s
- FLAC, true peak: 38.475s
- WAV, true peak: 36.283s

**Comment by diogocp** [comment]
I'm thinking it may make more sense to work on this on the Groove Basin side, though. Groove Basin can read all the metadata and then, based on availability/configuration/heuristics, decide whether to do loudness scanning with grooveloudness.

**Comment by andrewrk** [comment]
I'm doing some maintenance on libgroove right now - please re-open this issue on Groove Basin issue tracker if you want to discuss it further.
